applicationset/services/scm_provider/azure_devops.go

```diff
@@ -0,0 +1,157 @@
+
+type AzureDevopsProvider struct {
+	organization  string
+	teamProject   string
+	accessToken   string
+	clientFactory AzureDevopsClientFactory
+}
+
+var _ SCMProviderService = &AzureDevopsProvider{}
+var _ AzureDevopsClientFactory = &devopsFactoryImpl{}
+
+func NewAzureDevopsProvider(ctx context.Context, accessToken string, org string, project string) (*AzureDevopsProvider, error) {
+	if accessToken == "" {
+		return nil, fmt.Errorf("No access token provided")
+	}
+
+	baseUrl, err := getBaseUrl()
+	if err != nil {
+		return nil, err
+	}
+	url := fmt.Sprintf("%s/%s", *baseUrl, org)
+	connection := azuredevops.NewPatConnection(url, accessToken)
+	return &AzureDevopsProvider{organization: org, teamProject: project, accessToken: accessToken, clientFactory: &devopsFactoryImpl{connection: connection}}, nil
+}
+
+func (g *AzureDevopsProvider) ListRepos(ctx context.Context, cloneProtocol string) ([]*Repository, error) {
+	gitClient, err := g.clientFactory.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	getRepoArgs := azureGit.GetRepositoriesArgs{Project: &g.teamProject}
+	azureRepos, err := gitClient.GetRepositories(ctx, getRepoArgs)
+
+	if err != nil {
+		return nil, err
+	}
+	repos := []*Repository{}
+	for _, azureRepo := range *azureRepos {
+		if azureRepo.Name == nil || azureRepo.DefaultBranch == nil || azureRepo.RemoteUrl == nil || azureRepo.Id == nil {
+			continue
+		}
+		repos = append(repos, &Repository{
+			Organization: g.organization,
+			Repository:   *azureRepo.Name,
+			URL:          *azureRepo.RemoteUrl,
+			Branch:       *azureRepo.DefaultBranch,
+			Labels:       []string{},
+			RepositoryId: *azureRepo.Id,
+		})
+	}
+
+	return repos, nil
+}
+
+func (g *AzureDevopsProvider) RepoHasPath(ctx context.Context, repo *Repository, path string) (bool, error) {
+	gitClient, err := g.clientFactory.GetClient(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	idAsString := fmt.Sprintf("%v", repo.RepositoryId)
+	branchName := repo.Branch
+	getItemArgs := azureGit.GetItemArgs{RepositoryId: &idAsString, Project: &g.teamProject, Path: &path, VersionDescriptor: &azureGit.GitVersionDescriptor{Version: &branchName}}
+	_, err = gitClient.GetItem(ctx, getItemArgs)
+
+	if err != nil {
+		if wrappedError, isWrappedError := err.(azuredevops.WrappedError); isWrappedError {
+			if *wrappedError.TypeKey == "GitItemNotFoundException" {
+				return false, nil
+			}
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (g *AzureDevopsProvider) GetBranches(ctx context.Context, repo *Repository) ([]*Repository, error) {
+	gitClient, err := g.clientFactory.GetClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	getBranchesRequest := azureGit.GetBranchesArgs{RepositoryId: &repo.Repository, Project: &g.teamProject}
+	branches, err := gitClient.GetBranches(ctx, getBranchesRequest)
+	if err != nil {
+		return []*Repository{}, nil //Repo might be locked/unavailable, branch authz; all sorts of reasons why this would fail. Just return empty result.
```

Comment:
I'm a little hesitant to ignore this error. A network failure would cause the branches to be dropped and (potentially) Applications to be deleted. I think we should throw the error and then rely on the user to fix the underlying issues or contribute features to ignore the errors.

Reply:
Thanks, I completely agree. The code now does this: 

```
	if err != nil {
		return nil, fmt.Errorf("failed getting branches from repository %v, project %v: %w", repo.Repository, g.teamProject, err)
	}
```

Reply:
As Azure DevOps returns metadata on disabled repositories in its `GetRepositories` results, subsequent processing in `GetBranches`/`GetBranch` fails. 
I could not find any indications in the Azure DevOps Go API to determine if a repository is disabled.

When encountering errors from GetBranches, the code now checks for the specific TypeKey `GitRepositoryNotFoundException` and returns an empty result. 

As the inputs to `GetBranches` in Azure DevOps are values just received from the very same instance, I think the risk of app deletion is small, as the repo most likely was disabled when the initial set of repositories was returned in `ListRepos`. Any other error is wrapped and returned, which should ensure that most  error situations are signaled and rely on the user to investigate and fix.

I feel this was the best I could come up with, but any thoughts and inputs on this on the matter are welcome :)